src/lib/raop_ntp.c

```diff
@@ -0,0 +1,86 @@
+/**
+ *  Copyright (C) 2014  Joakim Plate
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ */
+
+#include "raop_ntp.h"
+#include <string.h>
+#include <stdlib.h>
+
+void
+raop_ntp_init(raop_ntp_t *ntp, logger_t *logger, unsigned long long clock)
+{
+	int i;
+	ntp->logger = logger;
+	for(i = 0; i < RAOP_NTP_COUNT; ++i) {
+		ntp->data[i].offset     = 0ll;
+		ntp->data[i].delay      = RAOP_NTP_MAXDISP;
+		ntp->data[i].dispersion = RAOP_NTP_MAXDISP;
+		ntp->data[i].clock      = clock;
+	}
+}
+
+static int
+raop_ntp_compare(const void* av, const void* bv)
+{
+	const raop_ntp_data_t* a = (const raop_ntp_data_t*)av;
+	const raop_ntp_data_t* b = (const raop_ntp_data_t*)bv;
+	if(a->delay < b->delay) {
+		return -1;
+	} else if(a->delay > b->delay) {
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+void
+raop_ntp_get_offset(raop_ntp_t *ntp, unsigned long long clock, long long *offset, unsigned long long *dispersion)
+{
+	raop_ntp_data_t data[RAOP_NTP_COUNT];
+	const unsigned  two_pow_n[RAOP_NTP_COUNT] = {2, 4, 8, 16, 32, 64, 128, 256};
+	int i;
+
+	/* sort by delay */
+	memcpy(data, ntp->data, sizeof(data));
+	qsort(data, RAOP_NTP_COUNT, sizeof(data[0]), raop_ntp_compare);
+
+	*dispersion = 0ull;
+	*offset     = ntp->data[0].offset;
```

Comment:
I think it should be data[0] instead of ntp->data[0] here. I'm aware this is an obsolete PR, but I found it useful as a reference for how synchronisation can be added.

Reply:
Hehe, Yea it was some time ago. Too bad we never got it finished. You looking at it for shairplay now?

Reply:
I'm looking at it for my Raspberry Pi AirPlay mirroring receiver RPiPlay (which is based on @dsafa22's AirplayServer, which is based on ShairPlay). I wonder why all this NTP stuff isn't in the ShairPlay master (anymore?).

Reply:
Not sure. Maybe experimental branch never got merged.